# Introduction

The goal of this project is to analyze non-normal distributed data of open source GitHub projects from a dataset provided by .

The project should show an ability to explore data, handle ambiguity, and weave insights into a narrative. It should have well-formed observations and understanding of the nuances of the data.

In [2]:
# Import required modules
import pandas as pd
import matplotlib.pyplot as plt
from scipy import stats
import numpy as np
import seaborn as sns
from datetime import timedelta

# Data Extraction

The dataset is provided from two files "usage.csv" and "projects.csv". They are selected usage records of open source GitHub projects.

In [5]:
usage = pd.read_csv("usage.csv")

# To view initial structure of dataset
usage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 566094 entries, 0 to 566093
Data columns (total 6 columns):
id            566094 non-null object
actor_id      444652 non-null object
project_id    566094 non-null object
account_id    566094 non-null object
started_at    566094 non-null object
ended_at      566094 non-null object
dtypes: object(6)
memory usage: 25.9+ MB


There are 566094 observations/rows in usage and 6 columns/variables. Every column, except for the actor_id has non-null values.

In [6]:
project = pd.read_csv("projects.csv")
project.columns=['project_id','project_size']
project.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12907 entries, 0 to 12906
Data columns (total 2 columns):
project_id      12907 non-null object
project_size    12906 non-null float64
dtypes: float64(1), object(1)
memory usage: 201.8+ KB


The initial view of the projects data file shows that there is 1 null value.

In [7]:
# Merging two datasets
merged = pd.merge(usage, project, on='project_id', how='outer')

In [8]:
merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 566094 entries, 0 to 566093
Data columns (total 7 columns):
id              566094 non-null object
actor_id        444652 non-null object
project_id      566094 non-null object
account_id      566094 non-null object
started_at      566094 non-null object
ended_at        566094 non-null object
project_size    563510 non-null float64
dtypes: float64(1), object(6)
memory usage: 34.6+ MB


# Data Wrangling/Munging

In [9]:
merged['started_at'] = pd.to_timedelta('00:' + merged['started_at'], errors='coerce') 
merged['ended_at'] = pd.to_timedelta('00:' + merged['ended_at'], errors='coerce')   

Due to the non-powerful nature of my tablet-laptop, I have subsetted the data to 384 observations. This number was determined by calculating the sample size with a population of 566,094 (the total number of observations), a confidence level of 95%, and a margin of error of 5%. Since I am limiting the dataset, I also removed all null values.

In [10]:
no_na = merged[pd.notnull(merged['actor_id'])]
no_na = no_na[pd.notnull(merged['project_size'])]
small = no_na.sample(384)
small = small.reset_index(drop=True)

C:\Users\Kevin\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


In [11]:
small.head()

id  \
0  a728cc5a62c4b5d6130f0662ac03c15f6ddce72f31eb2b...   
1  85406eda28c3da86d1d073b37e177daddf3fc56816acd1...   
2  4ca69276b6d9ae85ce07235bbdc7a00685d4754df39fd6...   
3  48a2f4b9710d464c4e9b1f39442bdb745e1b8be40fab30...   
4  cebfe80528039f7b25c90d2c9de20583b69bb72c61e1b0...   

                                            actor_id  \
0  cb03bf60e166ed64d3ec4a6730c023384e645e7180c33e...   
1  c3e2d9e519da2a615b9d83669116d822f0d30795cfe364...   
2  47acc019ec56ad25e6cd44ddbe2ea513702ceeb6a1c17a...   
3  467a3f755fafd35efc28ff5ebacd193c61c5e9ba388a84...   
4  6c83fc5ec41ead6ea3336a98e7d3614b98876a2972de11...   

                                          project_id  \
0  79543f158ad9c27582a6e5517ab5ef91deca2be002d5e9...   
1  2d0e24be7b4c7406c4efd0772f1351707eba06d201326f...   
2  878e1e82ff0e64486463fbe10886931cd60b67d70f4dfa...   
3  f09396375c9a706d291c6264cdcfd441182d9d422f2b7d...   
4  dba05c9d71b2b3fb1a3a94bf80cccdb7ecd4776cc63776...   

                                          account_id      started_at  \
0  f332c66eae839044f0476a26b83367426dbc04beab90aa... 00:48:16.500000   
1  16876b96b6b8b2ebc48c543c18080cdc8a84a87eb1c515... 00:49:10.600000   
2  892bf9181e5034cbc80b08ee0dd5598bdefa7dc4880b06...        00:56:24   
3  d2fefe1774b77114507fdce4a9c63cf13a6bede7afff9d...        00:37:03   
4  eeccd1b657ba995781fb315f165e9be7949dbadf88a42c... 00:20:57.100000   

         ended_at  project_size  
0        00:50:51       44441.0  
1        00:51:26       10383.0  
2 00:59:39.700000      212719.0  
3 00:38:02.900000          71.0  
4 00:22:28.300000         640.0

Since the imported data only measured time in minutes and dates, events such as the start time being after the end time occured. This is not possible. Therefore the next code segment checks that the laws of time have not been broken, if so, it makes adds an hour until the end date is after the start date. This results in a more accurate and realistic time of, for example 2 minutes, than 23 hours and 58 minutes.

In [12]:
i=0
while i < len(small):
    try:
        while small['ended_at'][i]<small['started_at'][i]:
            small['ended_at'][i] = small['ended_at'][i]+timedelta(0,60*60)
    except TypeError:
        pass
    i += 1

C:\Users\Kevin\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """


In [13]:
# Calculate time_of_usage

calc = small.apply(lambda row: row['ended_at'] - row['started_at'], axis = 1)

small['time_of_usage'] = calc

# Check / Test
small

id  \
0    a728cc5a62c4b5d6130f0662ac03c15f6ddce72f31eb2b...   
1    85406eda28c3da86d1d073b37e177daddf3fc56816acd1...   
2    4ca69276b6d9ae85ce07235bbdc7a00685d4754df39fd6...   
3    48a2f4b9710d464c4e9b1f39442bdb745e1b8be40fab30...   
4    cebfe80528039f7b25c90d2c9de20583b69bb72c61e1b0...   
5    7a565fa07b10d9f7a66984a4c9ed13c8c49267357a700f...   
6    b89d8170355802a3e00d0b656539d733a8276cf9ca7901...   
7    516750b3d59134b5122609431c8f2399120661aae8fc7f...   
8    5499eb08e9ab2fc1500c0c563185748676643726e1c989...   
9    7167d3f0a5bb647f45d908e421051876bfa201fcb6e2a9...   
10   987951dba8914033409308ef9cdff601b6e82c8cae350f...   
11   8fd93cc453be6bfc48de7cf08575742029ddc2fcdd4284...   
12   9f109922d5490a103864cdc5bb56841e2145341a1f1827...   
13   a621a9ddedeeb547410cba8acea2d0188d5e89eb174e9a...   
14   4ba405aff40b98cf415953a62c45a5bfe2bdde09c50870...   
15   d8276da4c797f5a65bbff86b1ba28b7bbd4b867889b081...   
16   c764a993b2758aea3b8563967dd497a5c7f5fa992e1f47...   
17   b42f900398d3f0ef91dfb4885f4f1f2a8a7dc3453967fc...   
18   4605256eb59f08bc592d92393c4adc703409e1730bb3e7...   
19   2b3ac9cf9d9d7bd0ae2586e7489adf0a1ea35be822d00d...   
20   686930d77d1410f88eb104776e018f596469d4ba11e9ed...   
21   354ca98b0ea7e24b0ce268dec13158f3a9260c2cf92792...   
22   c9aaa796e40676de1021355840670ffede5338f6a110f1...   
23   546326abe66d63879a15e6f7073fca31cfcf7717a2508a...   
24   f1149da9b673f5144c27485102dd13400bcb272a1daed2...   
25   c7876c337c12e3230d6d80db1b27f07c0858b4638b642a...   
26   489364e9cdab23dce2ec2dd7975f29004bf826d8ed4c4e...   
27   67eb2f0fe68c4b4ba54a0c6f7d7c365b04d0bdf0dc57ac...   
28   a83e29a81bf878671a275790efcf442421987bc047a127...   
29   27cacca88174201df04b1cc89d454f48ccf217af3ea2c9...   
..                                                 ...   
354  564438e1b23c88bfc8bd05fed26026f0ef7f2c9ca6f694...   
355  8195b2b73e47fb248d38b2f68bf2de0e5ea9847f8e7dd2...   
356  5f15362ac91a675893660154e5082f5dbf8831d1f30ae5...   
357  082d2feacee6e0e08fefb8d2e4876c051e66c1cde4ffd4...   
358  0fe6f0e4af8893060cf78815f1b1b5c7c9a56a96749795...   
359  4f1d468089f799c5b3bbaed24a2e4d03779e627375ba04...   
360  650395bbc2c31b5fdd2cc4d034506c211631f297ca317a...   
361  1dcb1fddbfc5622edf9be7ca8ec03a1d994747dce262eb...   
362  b77e929d2b42e80558416829dbacd16639b048b343755e...   
363  2d7de381c5e970e83ac52047b01c17726dcfd874f9327f...   
364  c28452a30c772136938c880af9d6b21483ee15fba9358c...   
365  3da26caecd899d639ca5224cd252c3eb604e7be82d8e7b...   
366  96093fd23ed0782f29b57a83c5b6dd3fcaaaeb4279e13c...   
367  3e5126e8519e50a154df33f0ebbb35dc68752ca5d7636e...   
368  ede9c755f5a35b2109bc06d48ee8d8a22b76399d8f56d0...   
369  7801737143eaff45619af7959752003d64088607ae13f6...   
370  49b98f8c8528f11083876c4ddb576b7dc8e58a50bec7ce...   
371  1d645838563578a05df57b378ff683e0d2d9cd46681456...   
372  4b9bdde6a51869096b885e7d86f6ad372a61c757801919...   
373  786b9658e5ab73034569fbac65f0df23b4d601b593cf9e...   
374  7403bc18afc1f9ffda5f9fee56705e8a0f2e5b69866ec2...   
375  92d91aff0145aea426da59c6bd890dc530b613f51799ef...   
376  1155388021a74e6fb415093fef1ccc844c8e4d017cc260...   
377  702b23562463f987ad66146149b724a7ced27f4f9ed3bd...   
378  cdec3b18d5e43fa75a132fc41893c2e142c14dbefc8b9b...   
379  b7e20fa96905bbf151ff1c770ce871c659f005bc1a03d9...   
380  9e041ef19b0bc418154c9c89b2f27285013a21feb6892d...   
381  5b31407fce99e49768553934ccb6cad2c41ea16236345c...   
382  6a166dc8b7c98876f53254ac4b48d0f432b8cd37e0d553...   
383  56016f1d734dfa3b7f3f54c5d26136a3432c529d9c3c4f...   

                                              actor_id  \
0    cb03bf60e166ed64d3ec4a6730c023384e645e7180c33e...   
1    c3e2d9e519da2a615b9d83669116d822f0d30795cfe364...   
2    47acc019ec56ad25e6cd44ddbe2ea513702ceeb6a1c17a...   
3    467a3f755fafd35efc28ff5ebacd193c61c5e9ba388a84...   
4    6c83fc5ec41ead6ea3336a98e7d3614b98876a2972de11...   
5    b6179cfed84f3a33c9263d3df38b85b30d12e8e49b0178...   
6    4e0ac40f438138ad8058fd83ccca4e78da87ff402c6a

#  Goals

#1 Identify the top 10 actors - by total usage, by accounts built against, and by projects built against including the scalar amounts

In [30]:
df = small
df.groupby('actor_id')

#.size().nlargest(10)

TypeError: 'DataFrameGroupBy' object is not callable

In [21]:
# By Acccounts Built Against

acc = pd.DataFrame(small.groupby(by=['actor_id'])['account_id'].value_counts())
acc.columns = ['scalar']
acc.sort_index(axis='scalar')



ValueError: No axis named scalar for object type <class 'pandas.core.frame.DataFrame'>

In [38]:
df = small
df_count = df.groupby('actor_id')
pd.DataFrame(df.groupby(by=['actor_id']))

ValueError: DataFrame constructor not properly called!

In [ ]:
# By Total Usage

In [16]:
# By Projects Built Against

#2 Provide a five-number summary of usage for each week, and a five-number summary of total usage per account for each week

In [ ]:
# The provided datasets do not specify weeks so it is not possible to know if these events
# happened on different weeks or the same week

merged.describe()

#3 Provide some comments on the following: Where can a t-test be applied to these collections? Provide some examples of where it can and cannot, insights on why, and other implications

In [ ]:
A t-test can be us

#4 Provide an analysis that shows the relationship between the following variables, along with insights drawn out and comments on your choices:
- Total usage per account
- Unique Actors per account
- Two other variables

In [ ]:
rho, pval = stats.spearmanr(small)

pvals = pd.DataFrame(pval)
#pvals = pvals.rename(index={0: 'outlook',1: 'net_income',2: 'ft_employee',3: 'div_yield',4: 'five_year',5: 'ten_year',6: 'alcohol',7: 'gambling',8: 'tobacco',9: 'controversial_weapons',10: 'small_arms',11: 'military_contracting',12: 'coal',13: 'dont_use',14: 'dont_trust',15: 'assets',16: 'market_value',17: 'revenue',18: 'net_per_emp'})
#round(pvals, 3)
pval

#5 Using these collections, determine what are the main drivers (if any) of usagegrowth for accounts.